rclcpp/include/rclcpp/timer.hpp

```diff
@@ -106,29 +95,49 @@ class GenericTimer : public TimerBase
   GenericTimer(std::chrono::nanoseconds period, CallbackType callback)
   : TimerBase(period, callback), loop_rate_(period)
   {
-    thread_ = std::thread(&GenericTimer<Clock>::run, this);
+    /* Subtracting the loop rate period ensures that the callback gets triggered
+       on the first call to check_and_trigger. */
+    last_triggered_time_ = Clock::now() - period;
   }
 
   virtual ~GenericTimer()
   {
     cancel();
-    thread_.join();
   }
 
-  void
-  run()
+  // return: true to trigger callback on the next "execute_timer" call in executor
+  bool
+  check_and_trigger()
+  {
+    if (canceled_) {
+      return false;
+    }
+    if (Clock::now() < last_triggered_time_) {
+      // last trigger time was in the future, this should never happen
+      throw std::runtime_error("Received invalid last triggered time");
+    }
+    if (std::chrono::duration_cast<std::chrono::nanoseconds>(Clock::now() - last_triggered_time_) >=
+      loop_rate_.period())
+    {
+      last_triggered_time_ = Clock::now();
+      return true;
+    }
+    return false;
+  }
+
+  std::chrono::nanoseconds
+  time_to_trigger()
```

Comment:
Minor comment, `time_to_trigger` sounds like the absolute (datetime) time on which to trigger. Maybe `time_until_trigger` or `time_till_trigger` better indicates that this is a duration until the next trigger.